tiledb/sm/serialization/query.cc

```diff
@@ -261,86 +261,106 @@ Status subarray_to_capnp(
   return Status::Ok();
 }
 
-Status subarray_from_capnp(
-    const capnp::Subarray::Reader& reader, Subarray* subarray) {
-  RETURN_NOT_OK(subarray->set_coalesce_ranges(reader.getCoalesceRanges()));
+Subarray subarray_from_capnp(
+    const capnp::Subarray::Reader& reader,
+    const Array* array,
+    Layout layout,
+    stats::Stats* parent_stats,
+    shared_ptr<Logger> logger) {
+  bool coalesce_ranges = reader.getCoalesceRanges();
   auto ranges_reader = reader.getRanges();
+
   uint32_t dim_num = ranges_reader.size();
+  std::vector<RangeSetAndSuperset> range_subset(dim_num);
+  std::vector<bool> is_default(dim_num, false);
   for (uint32_t i = 0; i < dim_num; i++) {
     auto range_reader = ranges_reader[i];
     Datatype type = Datatype::UINT8;
-    RETURN_NOT_OK(datatype_enum(range_reader.getType(), &type));
+    throw_if_not_ok(datatype_enum(range_reader.getType(), &type));
+    auto dim = array->array_schema_latest().dimension_ptr(i);
 
-    auto data = range_reader.getBuffer();
-    auto data_ptr = data.asBytes();
+    bool implicitly_initialized = range_reader.getHasDefaultRange();
+    range_subset[i] = RangeSetAndSuperset(
+        dim->type(), dim->domain(), implicitly_initialized, coalesce_ranges);
+    is_default[i] = implicitly_initialized;
     if (range_reader.hasBufferSizes()) {
       auto ranges = range_buffers_from_capnp(range_reader);
-      RETURN_NOT_OK(subarray->set_ranges_for_dim(i, ranges));
-
-      // Set default indicator
-      subarray->set_is_default(i, range_reader.getHasDefaultRange());
+      // If the range is implicitly initialized, the RangeSetAndSuperset
+      // constructor will initialize the ranges to the domain.
+      if (!implicitly_initialized) {
+        // Add custom ranges, clearing any implicit ranges previously set.
+        range_subset[i] = RangeSetAndSuperset(
+            dim->type(), dim->domain(), ranges, coalesce_ranges);
+      }
     } else {
       // Handle 1.7 style ranges where there is a single range with no sizes
+      auto data = range_reader.getBuffer();
+      auto data_ptr = data.asBytes();
       Range range(data_ptr.begin(), data.size());
-      RETURN_NOT_OK(subarray->set_ranges_for_dim(i, {range}));
-      subarray->set_is_default(i, range_reader.getHasDefaultRange());
+      throw_if_not_ok(range_subset[i].add_range_unrestricted(range));
     }
   }
 
+  std::vector<optional<Subarray::LabelRangeSubset>> label_range_subset(
```

Comment:
Same comment here for using reserve and emplace back if possible.

Reply:
We need the nullopts to fill in here for the dimensions that don't have label ranges set, or we segfault in checking for the optional values in checks like [Subarray::has_label_ranges](https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/subarray/subarray.cc#L1046). I went ahead and did this but if you're not a fan of the logic I can revert to using the move assignment. I just had to make sure we fill in with nullopts for all dimension indices without label ranges.